src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp

```diff
@@ -0,0 +1,310 @@
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_index)
+  {
+    const exprt &op0 = skipped.operands().at(0);
+    if(op0.type().id() == ID_array)
+    {
+      return op0;
+    }
+  }
+  return nullopt;
+}
+
+/// Returns `arr` if \p expr matches `&arr[idx]` where `arr` has an array type,
+/// `nullopt` otherwise.
+static optionalt<exprt> match_address_of_array_index(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_address_of)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    return match_array_index(subexpr);
+  }
+  return nullopt;
+}
+
+/// Returns `arr` if \p expr matches `*(&arr[idx])` where `arr` has an array
+/// type, `nullopt` otherwise.
+static optionalt<exprt> match_deref_of_address_of_array_index(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_dereference)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    return match_address_of_array_index(subexpr);
+  }
+  return nullopt;
+}
+
+/// Returns `arr` if \p expr matches `&arr[idx] + idx` where `arr` has an array
+/// type, `nullopt` otherwise.
+static optionalt<exprt>
+match_address_of_array_index_plus_offset(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_plus)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    return match_address_of_array_index(subexpr);
+  }
+  return nullopt;
+}
+
+/// Returns `arr` if \p expr matches `*(&arr[idx] + idx)` where `arr` has an
+/// array type, `nullopt` otherwise.
+static optionalt<exprt>
+match_deref_of_address_of_array_index_plus_offset(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_dereference)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    return match_address_of_array_index_plus_offset(subexpr);
+  }
+  return nullopt;
+}
+
+/// Returns `ptr` if \p expr matches `ptr + idx` where `ptr` has a pointer type,
+/// `nullopt` otherwise.
+static optionalt<exprt> match_ptr_plus_offset(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_plus)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    if(subexpr.type().id() == ID_pointer)
+      return subexpr;
+  }
+  return nullopt;
+}
+
+/// Returns `ptr` if \p expr matches `*(ptr + idx)` where `ptr` has a pointer
+/// type, `nullopt` otherwise.
+static optionalt<exprt> match_deref_of_ptr_plus_offset(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_dereference)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    return match_ptr_plus_offset(subexpr);
+  }
+  return nullopt;
+}
+
+/// Returns `lval` if \p expr matches pattern `&lval`, `nullopt` otherwise
+static optionalt<exprt> match_address_of(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_address_of)
+  {
+    return skipped.operands().at(0);
```

Comment:
As above: please avoid `.operands.at(0)` and use proper expression APIs instead.

Reply:
fixed